contrib/endpoints/src/api_manager/config_manager.cc

```diff
@@ -0,0 +1,181 @@
+  }
+}
+
+void ConfigManager::on_fetch_metadata(utils::Status status) {
+  if (!status.ok()) {
+    // We should not get here
+    global_context_->env()->LogError("Unexpected status: " + status.ToString());
+    config_roollout_callback_(utils::Status(Code::ABORTED, status.message()),
+                              empty_configs);
+    return;
+  }
+
+  // Update service_name
+  if (global_context_->service_name().empty()) {
+    global_context_->set_service_name(
+        global_context_->gce_metadata()->endpoints_service_name());
+  }
+
+  // Update config_id
+  if (global_context_->config_id().empty()) {
+    global_context_->config_id(
+        global_context_->gce_metadata()->endpoints_service_config_id());
+  }
+
+  // Call ApiManager call with Code::ABORTED, ESP will stop moving forward
+  if (global_context_->service_name().empty()) {
+    std::string msg = "API service name not specified in configuration files";
+    global_context_->env()->LogError(msg);
+    config_roollout_callback_(utils::Status(Code::ABORTED, msg), empty_configs);
+    return;
+  }
+
+  // TODO(jaebong) config_id should not be empty for the first version
+  // This part will be removed after the rollouts feature added
+  if (global_context_->config_id().empty()) {
+    std::string msg = "API config_id not specified in configuration files";
+    global_context_->env()->LogError(msg);
+    config_roollout_callback_(utils::Status(Code::ABORTED, msg), empty_configs);
+    return;
+  }
+
+  // Fetch service account token
+  GlobalFetchServiceAccountToken(global_context_, [this](utils::Status status) {
+    on_fetch_auth_token(status);
+  });
+}
+
+void ConfigManager::on_fetch_auth_token(utils::Status status) {
+  if (!status.ok()) {
+    // We should not get here
+    global_context_->env()->LogError("Unexpected status: " + status.ToString());
+    config_roollout_callback_(utils::Status(Code::ABORTED, status.message()),
+                              empty_configs);
+  }
+
+  // Fetch configs from the Inception
+  auto fetchInfo = std::make_shared<ConfigsFetchInfo>();
+
+  // For now, config manager has only one config_id (100% rollout)
+  fetchInfo->rollouts = {{global_context_->config_id(), 100}};
+  fetch_configs(fetchInfo);
+}
+
+// Fetch configs from rollouts. fetchInfo has rollouts and fetched configs
+void ConfigManager::fetch_configs(std::shared_ptr<ConfigsFetchInfo> fetchInfo) {
+  // Finished fetching configs.
+  if (fetchInfo->rollouts.size() <= (size_t)fetchInfo->index) {
+    // Failed to fetch all configs or rollouts are empty
+    if (fetchInfo->rollouts.size() == 0 || fetchInfo->configs.size() == 0) {
+      // first time, call the ApiManager callback function with an error
+      config_roollout_callback_(
+          utils::Status(Code::ABORTED, "Failed to load configs"),
+          empty_configs);
+      return;
+    }
+
+    // Update ApiManager
+    config_roollout_callback_(utils::Status::OK, fetchInfo->configs);
+    return;
```

Comment:
Fixed typo.

I need to add more logic into this block to support managed rollouts. 
For instance, handling rollout refresh and partial fetch.
I think logic needs to stay in the ConfigManager to avoid adding complicated logic